### PR TITLE
Add template for mpi_cmd_prefix

### PR DESCRIPTION
**Pull request by ocaisa**
Replaces #3259

### CODE REVIEW

File: test/framework/toolchain.py [hunk 1]
```diff
@@ -948,6 +948,44 @@ def test_nosuchtoolchain(self):
         tc = self.get_toolchain('intel', version='1970.01')
         self.assertErrorRegex(EasyBuildError, "No module found for toolchain", tc.prepare)
 
+    def test_mpi_cmd_prefix(self):
+        """Test mpi_exec_nranks function."""
+        self.modtool.prepend_module_path(self.test_prefix)
+
+        tc = self.get_toolchain('gompi', version='2018a')
+        tc.prepare()
+        self.assertEqual(tc.mpi_cmd_prefix(nr_ranks=2), "mpirun -n 2 ")
```
Comment (boegel) [inline test/framework/toolchain.py]: we should strip off the trailing whitespace imho, to avoid duplicate spaces when using something like:

```python
sanity_check_commands = ["%(mpi_cmd_prefix)s example"]

File: test/framework/robot.py [hunk 2]
```diff
@@ -424,15 +424,15 @@ def test_resolve_dependencies_minimal(self):
...
+            # (it doesn't make much sense, but it serves the purpose)
+            "   ('OpenMPI', '%(version_minor)s.1.2'),",  # available with GCC/6.4.0-2.28
             "   ('OpenBLAS', '0.2.20'),",  # available with GCC/6.4.0-2.28
             "   ('ScaLAPACK', '2.0.2', '-OpenBLAS-0.2.20'),",  # available with gompi/2018a
             "   ('SQLite', '3.8.10.2'),",
             "]",
             # toolchain as list line, for easy modification later;
-            # the use of %(version_minor)s here is mainly to check if templates are being handled correctly
-            # (it doesn't make much sense, but it serves the purpose)
-            "toolchain = {'name': 'foss', 'version': '%(version_minor)s018a'}",
```
[the remark this answers is not in the document]
Comment (ocaisa) [reply test/framework/robot.py:435]: This test was failing when this template was left in place, it wasn't clear to me where but my guess was that the new template triggers a version check of the toolchain somewhere which is now taking place before the templates have been resolved.

[the remark this answers is not in the document]
Comment (ocaisa) [reply test/framework/robot.py:435]: See https://github.com/easybuilders/easybuild-framework/runs/545312074 for example

[the remark this answers is not in the document]
Comment (ocaisa) [reply test/framework/robot.py:435]: Even after looking into it again, I still can't figure out *why* it happens, the error is:
```
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/home/alanc/EasyBuild_Git/easybuild-framework/test/framework/robot.py", line 457, in test_resolve_dependencies_minimal
    res = resolve_dependencies([bar], self.modtool)
  File "/home/alanc/EasyBuild_Git/easybuild-framework/easybuild/tools/robot.py", line 460, in resolve_dependencies
    raise_error_missing_deps(totally_missing, extra_msg="no easyconfig file or existing module found")
  File "/home/alanc/EasyBuild_Git/easybuild-framework/easybuild/tools/robot.py", line 323, in raise_error_missing_deps
    raise EasyBuildError(error_msg)
EasyBuildError: 'Missing dependencies: foss/%(version_minor)s018a (no easyconfig file or existing module found)'
```
but the dict of the easyconfig is perfectly fine with templates resolved.

[the remark this answers is not in the document]
Comment (ocaisa) [reply test/framework/robot.py:435]: @boegel I've spent an hour trying to figure out why but I cannot solve this one

[the remark this answers is not in the document]
Comment (boegel) [reply test/framework/robot.py:435]: OK, here's what's going on...

The first time that the template dict is being composed (with the first call to `EasyConfig._generate_template_values`), `self.template_values` is still entirely blank.

But when `self.toolchain` is being passed to `template_constant_dict`, we're creating a toolchain instance, which requires the easyconfig file for the toolchain itself. So the toolchain version can't include template values like `%(version_minor)s` anymore now, since that template can't be resolved when `self.toolchain` is first used...

Strictly speaking that's a regression, but using `%(version_minor)s` to define the toolchain version is a pretty strange use case, so I don't consider this to be a problem.

Long story short: this test was doing something pretty strange, and the change you made to move it to `dependencies` makes sense.

This test was added in the context of a fix for the use of templates in `dependencies` (see #1812), so the changed test should still cover that use case.

[the remark this answers is not in the document]
Comment (boegel) [reply test/framework/robot.py:435]: Actually, there's a clean fix for it, implemented in e763c3040...
